packages/prisma-client/prisma/migrations/20230106000143_tree_styles_data/migration.ts

```diff
@@ -0,0 +1,100 @@
+import { PrismaClient } from "./client";
+
+type StylesItem = {
+  breakpointId: string;
+  instanceId: string;
+  property: string;
+  value: unknown;
+};
+
+type Styles = StylesItem[];
+
+type CssRule = {
+  breakpoint: undefined | string;
+  style: Record<string, unknown>;
+};
+
+type Text = {
+  type: "text";
+  value: string;
+};
+
+type Instance = {
+  type: "instance";
+  id: string;
+  component: string;
+  cssRules: Array<CssRule>;
+  children: Array<Instance | Text>;
+};
+
+const convertTree = (instance: Instance, styles: Styles) => {
+  for (const rule of instance.cssRules) {
+    if (rule.breakpoint === undefined) {
+      continue;
+    }
+    for (const [property, value] of Object.entries(rule.style)) {
+      styles.push({
+        breakpointId: rule.breakpoint,
+        instanceId: instance.id,
+        property,
+        value,
+      });
+    }
+  }
+  instance.cssRules = [];
+  for (const child of instance.children) {
+    if (child.type === "instance") {
+      convertTree(child, styles);
+    }
+  }
+};
+
+export default async () => {
+  const client = new PrismaClient({
+    // Uncomment to see the queries in console as the migration runs
+    // log: ["query", "info", "warn", "error"],
+  });
+
+  await client.$transaction(
+    async (prisma) => {
+      let cursor: undefined | string = undefined;
+      let hasNext = true;
+
+      while (hasNext) {
+        const chunkSize = 1000;
+        const trees = await prisma.tree.findMany({
+          take: chunkSize,
+          orderBy: {
+            id: "asc",
+          },
+          ...(cursor
+            ? {
+                skip: 1, // Skip the cursor
+                cursor: { id: cursor },
+              }
+            : null),
+        });
+        cursor = trees.at(-1)?.id;
```

Comment:
wooow, didn't know this method was added to array to explicitly allow negative indexes NICE!

Reply:
Two primary reasons: negative access and type safety